File: packages/ios-swiftui/Sources/FarmerChatSwiftUI/Crash/CrashBridge.swift
import Foundation
import os

/// Detects the host app's crash reporting provider at runtime via `NSClassFromString`
/// and forwards SDK errors through it.
///
/// Supports Firebase Crashlytics, Sentry, and Bugsnag. Falls back gracefully
/// to no-op when no provider is present.
///
/// All methods guard defensively -- this class must NEVER crash the host app.
internal final class CrashBridge {

    private static let logger = Logger(
        subsystem: "org.digitalgreen.farmerchat",
        category: "CrashBridge"
    )

    /// Detected crash provider in the host app.
    enum CrashProvider: String {
        case firebase = "Firebase"
        case sentry = "Sentry"
        case bugsnag = "Bugsnag"
        case none = "None"
    }

    /// The currently detected crash provider.
    private(set) var provider: CrashProvider = .none

    /// Detect the crash provider available in the host app's runtime.
    ///
    /// Call once during SDK initialization. Checks for Firebase Crashlytics,
    /// Sentry, and Bugsnag in that priority order.
    func detect() {
        if NSClassFromString("FIRCrashlytics") != nil {
            provider = .firebase
        } else if NSClassFromString("SentrySDK") != nil {
            provider = .sentry
        } else if NSClassFromString("Bugsnag") != nil {
            provider = .bugsnag
        } else {
            provider = .none
        }
        Self.logger.debug("Detected crash provider: \(self.provider.rawValue)")
    }

    /// Report an error to the host app's crash provider.
    ///
    /// - Parameters:
    ///   - error: The error to report.
    ///   - breadcrumbs: Optional context breadcrumbs attached before reporting.
    func reportError(_ error: Error, breadcrumbs: [String] = []) {
        // Attach breadcrumbs first
        for crumb in breadcrumbs {
            addBreadcrumb(crumb)
        }

        switch provider {
        case .firebase:
            reportFirebase(error)
        case .sentry:
            reportSentry(error)
        case .bugsnag:
            reportBugsnag(error)
        case .none:
            Self.logger.warning("SDK error (no crash provider): \(error.localizedDescription)")
        }
    }

    /// Add a breadcrumb for crash context.
    ///
    /// - Parameter message: Descriptive breadcrumb text.
    func addBreadcrumb(_ message: String) {
        switch provider {
        case .firebase:
            // FIRCrashlytics.crashlytics().log(message)
            guard let clazz = NSClassFromString("FIRCrashlytics") as? NSObject.Type,
                  clazz.responds(to: NSSelectorFromString("crashlytics")),
                  let instance = clazz.perform(NSSelectorFromString("crashlytics"))?.takeUnretainedValue(),
                  instance.responds(to: NSSelectorFromString("log:")) else { return }
            _ = instance.perform(NSSelectorFromString("log:"), with: message)

        case .sentry:
            // Create a SentryBreadcrumb and add it via SentrySDK.addBreadcrumb:
            guard let sentryClass = NSClassFromString("SentrySDK"),
                  let breadcrumbClass = NSClassFromString("SentryBreadcrumb") as? NSObject.Type else { return }
            let breadcrumb = breadcrumbClass.init()
            if breadcrumb.responds(to: NSSelectorFromString("setMessage:")) {
                _ = breadcrumb.perform(NSSelectorFromString("setMessage:"), with: message)
            }
            if breadcrumb.responds(to: NSSelectorFromString("setLevel:")) {
                // SentryLevel.info = 1
                _ = breadcrumb.perform(NSSelectorFromString("setLevel:"), with: NSNumber(value: 1))
            }
            let addSelector = NSSelectorFromString("addBreadcrumb:")
            if (sentryClass as AnyObject).responds(to: addSelector) {
                _ = (sentryClass as AnyObject).perform(addSelector, with: breadcrumb)
            }

        case .bugsnag:
            // Bugsnag.leaveBreadcrumb(withMessage:)
            guard let clazz = NSClassFromString("Bugsnag") else { return }
            let selector = NSSelectorFromString("leaveBreadcrumbWithMessage:")
            guard (clazz as AnyObject).responds(to: selector) else { return }
            _ = (clazz as AnyObject).perform(selector, with: message)

        case .none:
            break
        }
    }

    /// Set a custom key-value pair on crash reports for additional SDK context.
    ///
    /// - Parameters:
    ///   - key: The key name.
    ///   - value: The value.
    func setCustomKey(_ key: String, value: String) {
        switch provider {
        case .firebase:
            // FIRCrashlytics.crashlytics().setCustomValue(value, forKey: key)
            guard let clazz = NSClassFromString("FIRCrashlytics") as? NSObject.Type,
                  clazz.responds(to: NSSelectorFromString("crashlytics")),
                  let instance = clazz.perform(NSSelectorFromString("crashlytics"))?.takeUnretainedValue() else { return }
            let selector = NSSelectorFromString("setCustomValue:forKey:")
            guard instance.responds(to: selector) else { return }
            _ = instance.perform(selector, with: value, with: key)

        case .sentry:
            // Sentry's Swift-only API does not expose a simple setTag via ObjC selectors.
            // configureScope requires a closure which cannot be passed via perform:with:.
            // This is a known limitation of the reflection-based approach.
            Self.logger.debug("Sentry setCustomKey not available via ObjC reflection (key: \(key))")

        case .bugsnag:
            // Bugsnag.addMetadata(_:toSection:) with a dictionary
            // NSObject.perform only supports up to 2 arguments, so pass a dict.
            guard let clazz = NSClassFromString("Bugsnag") else { return }
            let selector = NSSelectorFromString("addMetadata:toSection:")
            guard (clazz as AnyObject).responds(to: selector) else { return }
            let metadata: NSDictionary = [key: value]
            _ = (clazz as AnyObject).perform(selector, with: metadata, with: "farmerchat")

        case .none:
            break
        }
    }

    // MARK: - Private Helpers

    /// Report error to Firebase Crashlytics via `recordError:`.
    private func reportFirebase(_ error: Error) {
        guard let clazz = NSClassFromString("FIRCrashlytics") as? NSObject.Type,
              clazz.responds(to: NSSelectorFromString("crashlytics")),
              let instance = clazz.perform(NSSelectorFromString("crashlytics"))?.takeUnretainedValue() else {
            return
        }
        let selector = NSSelectorFromString("recordError:")
        guard instance.responds(to: selector) else { return }
        _ = instance.perform(selector, with: error as NSError)
    }

    /// Report error to Sentry via `captureError:`.
    private func reportSentry(_ error: Error) {
        guard let sentryClass = NSClassFromString("SentrySDK") else { return }
        let selector = NSSelectorFromString("captureError:")
        guard (sentryClass as AnyObject).responds(to: selector) else { return }
        _ = (sentryClass as AnyObject).perform(selector, with: error as NSError)
    }

    /// Report error to Bugsnag via `notifyError:`.
    private func reportBugsnag(_ error: Error) {
        guard let clazz = NSClassFromString("Bugsnag") else { return }
        let selector = NSSelectorFromString("notifyError:")
        guard (clazz as AnyObject).responds(to: selector) else { return }
        _ = (clazz as AnyObject).perform(selector, with: error as NSError)
    }
}

File: packages/android-compose/src/main/kotlin/org/digitalgreen/farmerchat/compose/crash/CrashBridge.kt
package org.digitalgreen.farmerchat.compose.crash

import android.util.Log

/**
 * Detects the host app's crash reporting provider at runtime via reflection
 * and forwards SDK errors through it.
 *
 * Supports Firebase Crashlytics, Sentry, and Bugsnag. Falls back gracefully
 * to no-op when no provider is present.
 *
 * All public methods are wrapped in try-catch — this class must NEVER crash the host app.
 */
internal class CrashBridge {

    private companion object {
        const val TAG = "FC.CrashBridge"
    }

    /** Detected crash provider in the host app. */
    enum class CrashProvider { FIREBASE, SENTRY, BUGSNAG, NONE }

    private var provider: CrashProvider = CrashProvider.NONE

    /**
     * Detect the crash provider available in the host app's classpath.
     * Call once during SDK initialization.
     */
    fun detect() {
        try {
            provider = when {
                classExists("com.google.firebase.crashlytics.FirebaseCrashlytics") ->
                    CrashProvider.FIREBASE
                classExists("io.sentry.Sentry") ->
                    CrashProvider.SENTRY
                classExists("com.bugsnag.android.Bugsnag") ->
                    CrashProvider.BUGSNAG
                else ->
                    CrashProvider.NONE
            }
            Log.d(TAG, "Detected crash provider: $provider")
        } catch (e: Exception) {
            Log.w(TAG, "Failed to detect crash provider", e)
            provider = CrashProvider.NONE
        }
    }

    /** Returns the currently detected [CrashProvider]. */
    fun detectedProvider(): CrashProvider = provider

    /**
     * Report an error to the host app's crash provider.
     *
     * @param error The throwable to report.
     * @param breadcrumbs Optional context breadcrumbs attached before reporting.
     */
    fun reportError(error: Throwable, breadcrumbs: List<String> = emptyList()) {
        try {
            // Attach breadcrumbs first
            for (crumb in breadcrumbs) {
                addBreadcrumb(crumb)
            }

            when (provider) {
                CrashProvider.FIREBASE -> reportFirebase(error)
                CrashProvider.SENTRY -> reportSentry(error)
                CrashProvider.BUGSNAG -> reportBugsnag(error)
                CrashProvider.NONE -> Log.w(TAG, "SDK error (no crash provider)", error)
            }
        } catch (e: Exception) {
            // Never crash the host app — swallow and log
            Log.w(TAG, "Failed to report error to $provider", e)
        }
    }

    /**
     * Add a breadcrumb for crash context.
     *
     * @param message Descriptive breadcrumb text.
     */
    fun addBreadcrumb(message: String) {
        try {
            when (provider) {
                CrashProvider.FIREBASE -> {
                    val clazz = Class.forName("com.google.firebase.crashlytics.FirebaseCrashlytics")
                    val instance = clazz.getMethod("getInstance").invoke(null)
                    clazz.getMethod("log", String::class.java).invoke(instance, message)
                }
                CrashProvider.SENTRY -> {
                    val breadcrumbClass = Class.forName("io.sentry.Breadcrumb")
                    val breadcrumb = breadcrumbClass
                        .getMethod("info", String::class.java)
                        .invoke(null, message)
                    val sentryClass = Class.forName("io.sentry.Sentry")
                    sentryClass.getMethod("addBreadcrumb", breadcrumbClass)
                        .invoke(null, breadcrumb)
                }
                CrashProvider.BUGSNAG -> {
                    val clazz = Class.forName("com.bugsnag.android.Bugsnag")
                    clazz.getMethod("leaveBreadcrumb", String::class.java)
                        .invoke(null, message)
                }
                CrashProvider.NONE -> { /* no-op */ }
            }
        } catch (e: Exception) {
            Log.w(TAG, "Failed to add breadcrumb to $provider", e)
        }
    }

    /**
     * Set a custom key-value pair on crash reports for additional SDK context.
     *
     * @param key The key name.
     * @param value The value.
     */
    fun setCustomKey(key: String, value: String) {
        try {
            when (provider) {
                CrashProvider.FIREBASE -> {
                    val clazz = Class.forName("com.google.firebase.crashlytics.FirebaseCrashlytics")
                    val instance = clazz.getMethod("getInstance").invoke(null)
                    clazz.getMethod("setCustomKey", String::class.java, String::class.java)
                        .invoke(instance, key, value)
                }
                CrashProvider.SENTRY -> {
                    val sentryClass = Class.forName("io.sentry.Sentry")
                    // Sentry uses configureScope with a lambda, but we can use setTag directly
                    // via Sentry.configureScope. Simplest reflection path: setTag on HubAdapter.
                    // Alternative: Use Sentry.setTag which is a static convenience method.
                    sentryClass.getMethod("setTag", String::class.java, String::class.java)
                        .invoke(null, key, value)
                }
                CrashProvider.BUGSNAG -> {
                    val clazz = Class.forName("com.bugsnag.android.Bugsnag")
                    clazz.getMethod("addMetadata", String::class.java, String::class.java, Any::class.java)
                        .invoke(null, "farmerchat", key, value)
                }
                CrashProvider.NONE -> { /* no-op */ }
            }
        } catch (e: Exception) {
            Log.w(TAG, "Failed to set custom key on $provider", e)
        }
    }

    // ── Private helpers ──────────────────────────────────────────────

    private fun reportFirebase(error: Throwable) {
        val clazz = Class.forName("com.google.firebase.crashlytics.FirebaseCrashlytics")
        val instance = clazz.getMethod("getInstance").invoke(null)
        clazz.getMethod("recordException", Throwable::class.java)
            .invoke(instance, error)
    }

    private fun reportSentry(error: Throwable) {
        val clazz = Class.forName("io.sentry.Sentry")
        clazz.getMethod("captureException", Throwable::class.java)
            .invoke(null, error)
    }

    private fun reportBugsnag(error: Throwable) {
        val clazz = Class.forName("com.bugsnag.android.Bugsnag")
        clazz.getMethod("notify", Throwable::class.java)
            .invoke(null, error)
    }

    private fun classExists(name: String): Boolean = try {
        Class.forName(name)
        true
    } catch (_: ClassNotFoundException) {
        false
    }
}
